evolver/app/routers/hardware.py

```diff
@@ -93,8 +93,15 @@ def get_calibrator_actions(hardware_name: str, request: Request):
     if not calibrator:
         raise CalibratorNotFoundError
     calibration_procedure = calibrator.calibration_procedure
+
     actions = [
-        {"name": action.name, "description": action.description} for action in calibration_procedure.get_actions()
+        {
+            "name": action.name,
+            "description": action.description,
+            "requires_input": action.requires_input,
```

Comment:
good idea, i latest revision has implemented this.

Reply:
See https://github.com/ssec-jhu/evolver-ng/blob/main/evolver/base.py#L33-L36

Pydantic doesn't (or at least didn't) specifically have a single meta on "required". If you inspect their code they check against:
``required = model_fields[field].default != pydantic_core.PydanticUndefined and model_fields[field].default_factory is not None``.

It would be best to add a trivial helper func to ``evolver.util`` to contain this that gets passed a model and field. I would also add a helper ontop of that that takes just a model and returns all required fields (that calls the former func). I'm happy to quickly add this if you'd like?

Reply:
This is to indicate whether the action requires any kind of input, not whether an input on the FormModel is requires a value...

for individual fields looks like pydantic is behaving as expected, e.g. in this screenshot since no default exists on the FormModel for this action "temperature" is marked as a required field (screenshot from the http response).
<img width="1150" alt="Screenshot 2024-12-02 at 12 07 15 PM" src="https://github.com/user-attachments/assets/f06f55bd-2e37-4db6-8005-b9fded4c65d3">

Reply:
ah i think i see what you mean "what if a FormModel consists entirely of optional fields?".

Reply:
maybe best to think of "requires_input" in this "meta" level as you put it, to mean "requires a form to be rendered".

Happy to rename if there's some different naming that's better.

Reply:
Exactly that sort of question. It's a narrow feature but trivial enough to code up and add, such that "required" semantics match pydantic model semantics. This gives the user/designer the option of adding form fields that can be tweaked but don't require user input thus given a little extra versatility to the framework.

Reply:
Though if this requires input is used to ensure a form is presented, then I think it probably doesn't even matter whether there are defaults/optional fields, the user should be presented with the form (defaults would just be filled out), and presumably always send the form back.

Reply:
yeah, frontend is always gonna render a form if this is true, and the actual details of that form are handled by pydantic and the thing parsing the pydantic defined form on the frontend (see screenshot).

Reply:
>requires a form to be rendered.

Then @amitschang suggestion works best.

However, I think if we might be doing this the wrong way around, no? This is a pydantic model, so shouldn't we just be returning the entire model, and then it's the UIs responsibility to determine what to do with it, e.g., based on whether the model is empty?

Reply:
From the frontend POV - we need someway to distinguish actions that are regular steps like "wait for ambient temp to equilibriate" and those that require user to manually input values.

This could be done by understanding that a `form_model: {}` implicitly indicates that no user input is required for this action, and writing logic on the front end to codify that (without any meaningful guarantees from the backend that that assumption will remain true) - or we can make a nice ergonomic API that explicitly states, "this action requires input, parse and render the form_model".

Up to you guys. just lmk what the HTTP api is that you envision and I'll implement. From the discussion so far it's not clear to me.

Reply:
Just removed "requires_input" entirely, frontend will take care of it.